### PR TITLE
  feat(components): InputText Automatic Resize

### PR DESCRIPTION
**Pull request by tyler-santos**
## Motivations
We need to be able to have a textarea that would start as 1 line tall, and grow to a set amount, and then become scrollable if more room was needed beyond that set maximum. 

The current implementation almost allowed us to do this, however it wouldn't allow us to set a minimum number of rows.

## Changes

Instead of just rows as a single number being passed into the InputText component, now we have the ability to pass in an associative array which has a min and max key, corresponding to the minimum and maximum number of rows that would like to be shown.

### Added

handleChange and calculateRows functions which actively monitor and calculates how many rows should be shown for a multiline text area **with** a set maximum and minimum number of rows.

### Changed

The only change in existing functionality is the removal of the min-height for text area so that we could have a minimum value less than 3. Instead we set the default rows value to be 3 for multiline, which can be overrided with any value. Essentially it behaves the exact same as before but now you can have less than 3 rows in the textarea for use cases such as ours.

### Deprecated

- 

### Removed

- 

### Fixed

- 

### Security

- 

## Testing

Added a jest test to ensure that the calculations were correct.

**QA**
From the input text page in Atlantis, try using the following code snippet in one of the testing areas.
`<InputText defaultValue="Rocinante" multiline rows={{min:1, max:6}} />`

Play with the min and max sizes to see different behaviours. 
Test to make sure that the behaviour of Rows as a single number still works and multiline still works as a standalone with a default number of visible rows of 3.

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/components/src/InputText/InputText.tsx [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
+  Ref,
+  createRef,
+  forwardRef,
+  useImperativeHandle,
+  useState,
+} from "react";
 import { XOR } from "ts-xor";
 import { FormField, FormFieldProps } from "../FormField";
 
+interface MultiRows {
```
Comment (JeffMarvin) [inline packages/components/src/InputText/InputText.tsx:11]: `RowRange` might be more accurate and clear for what this is doing.

File: packages/components/src/InputText/InputText.tsx [hunk 2]
```diff
@@ -50,14 +61,54 @@ function InputTextInternal(
...
     />
   );
 
+  function initializeRows() {
+    if (!props.rows) return;
+    if (typeof props.rows === "number") {
+      return props.rows;
+    } else {
+      return props.rows.min;
+    }
```
Comment (seanhealy) [inline packages/components/src/InputText/InputText.tsx:82]: I _think_ you might be able to simplify this with:

```suggestion
    if (typeof props.rows === "object") {
      return props.rows.min;
    }
    return props.rows;

```

File: packages/components/src/InputText/InputText.tsx [hunk 3]
```diff
@@ -50,14 +61,54 @@ function InputTextInternal(
...
+  }
+
+  function handleChange(newValue: string) {
+    props.onChange && props.onChange(newValue);
+    if (!props.rows) return;
+    if (
+      typeof props.rows !== "number" &&
+      inputRef &&
+      inputRef.current instanceof HTMLTextAreaElement
+    ) {
```
Comment (seanhealy) [inline packages/components/src/InputText/InputText.tsx:92]: Perhaps?

```suggestion
    if (
      props.rows &&
      typeof props.rows !== "number" &&
      inputRef &&
      inputRef.current instanceof HTMLTextAreaElement
    ) {
```

File: packages/components/src/InputText/InputText.tsx [hunk 4]
```diff
@@ -27,7 +38,7 @@ interface MultilineProps extends BaseProps {
   /**
    * Specifies the visible height of a long answer form field.
```
Comment (JeffMarvin) [inline packages/components/src/InputText/InputText.tsx:39]: Probably should update the comment to include the new option for this prop here.

File: packages/components/src/InputText/InputText.tsx [hunk 5]
```diff
@@ -50,14 +61,54 @@ function InputTextInternal(
...
+      const calculatedRows = calculateRows(
+        textareaScrollHeight,
+        textareaLineHeight,
+      );
+      if (
+        calculatedRows >= props.rows.min &&
+        calculatedRows <= props.rows.max
+      ) {
+        setRows(calculatedRows);
+      }
```
Comment (JeffMarvin) [inline packages/components/src/InputText/InputText.tsx:108]: This whole block could be a separate function. If you define it on line 111 (i.e. within the component scope) then it will have access to all of the scope that it currently does, so it shouldn't be too complicated.

Comment (JeffMarvin) [reply packages/components/src/InputText/InputText.tsx:108]: Basically, you could pass in `inputRef` and return `calculatedRows` which you would then use in `setRows(calculatedRows)`.
Happy to pair on that change if you want!

**Comment by JeffMarvin** [comment]
@tyler-santos Does this work? I am trying to get the rows to show up on some of these examples but I can't consistently get the right behaviour.  Can you post a quick QA plan for one way to show that this is working? 
https://deploy-preview-303--jobber-atlantis.netlify.app/components/input-text

**Comment by JeffMarvin** [comment]
@tyler-santos Oh, also, I think we should add an example block in the relevant Readme file for the component that shows how this new prop is used.

**Comment by tyler-santos** [comment]
> @tyler-santos Does this work? I am trying to get the rows to show up on some of these examples but I can't consistently get the right behaviour. Can you post a quick QA plan for one way to show that this is working?
> https://deploy-preview-303--jobber-atlantis.netlify.app/components/input-text

Yes it does work! Here is a little gif showing the behaviour. I will put in QA steps. 
Screen Recording 2020-07-10 at 2 19 42 PM

File: packages/components/src/InputText/InputText.tsx (hunk 3 above)
Comment (tyler-santos) [reply packages/components/src/InputText/InputText.tsx:92]: I think this would work as well but linter complains if I do it this way.

File: packages/components/src/InputText/InputText.tsx (hunk 4 above)
Comment (tyler-santos) [reply packages/components/src/InputText/InputText.tsx:39]: Good call thanks!

File: packages/components/src/InputText/InputText.tsx (hunk 5 above)
Comment (tyler-santos) [reply packages/components/src/InputText/InputText.tsx:108]: Chatted about this separately and got it resolved :)

**Comment by seanhealy** [comment]
@tyler-santos Just gave it a test it looks like it grows upon adding text / lines but isn’t shrinking if that text is removed.

**Comment by tyler-santos** [comment]
@GetJobber/frends I have requested a review on this, the old specs still need to be removed and new ones added but it would be good to get some feedback on the new implementation. Thank you very much!

**Comment by crpahl** [comment]
Tyler is out today so I'll be paying attention to any requested changes and addressing them for now 😄

File: .gitignore [hunk 6]
```diff
@@ -6,3 +6,4 @@ node_modules
 
 **/tsconfig.tsbuildinfo
 .DS_Store
+.idea
```
Comment (crpahl) [inline .gitignore:9]: I added this because I'm a dirty RubyMine user and it kept picking up changes to this RubyMine file.

File: packages/components/src/FormField/FormField.tsx [hunk 7]
```diff
@@ -172,7 +172,7 @@ export const FormField = React.forwardRef(
       onValidation,
       placeholder,
       readonly,
-      rows,
+      rows = 3,
```
Comment (crpahl) [inline packages/components/src/FormField/FormField.tsx:175]: Hmm is this change still needed?

I'm handling the default in `InputText.tsx`

```
  function getRowRange(): RowRange {
    if (props.rows === undefined) {
      return { min: 3, max: 3 };
    } else if (typeof props.rows === "object") {
      return { min: props.rows.min, max: props.rows.max };
    } else {
      return { min: props.rows, max: props.rows };
    }
  }
```

File: packages/components/src/InputText/InputText.tsx [hunk 8]
```diff
@@ -52,12 +61,53 @@ function InputTextInternal(
...
+      return { min: 3, max: 3 };
+    } else if (typeof props.rows === "object") {
+      return { min: props.rows.min, max: props.rows.max };
+    } else {
+      return { min: props.rows, max: props.rows };
+    }
+  }
+
+  function resize(textArea: HTMLTextAreaElement) {
+    if (rowRange.min === rowRange.max) return;
```
Comment (crpahl) [inline packages/components/src/InputText/InputText.tsx:91]: Return if no resizing is needed. This will handle the case where the values of rows is:

```
rows = {3}
```

or something like:

```
rows = {{min: 3, max: 3}}
```

File: packages/components/src/InputText/InputText.tsx [hunk 9]
```diff
@@ -52,12 +61,53 @@ function InputTextInternal(
...
+  function textAreaHeight(textArea: HTMLTextAreaElement) {
+    const maxHeight = rowRange.max * getLineHeight(textArea);
+    return Math.min(textArea.scrollHeight, maxHeight);
+  }
+
+  function getLineHeight(textArea: HTMLTextAreaElement) {
+    const { fontSize, lineHeight } = window.getComputedStyle(textArea);
+    const lh =
+      lineHeight === "normal"
+        ? parseFloat(fontSize) * 1.2
```
Comment (crpahl) [inline packages/components/src/InputText/InputText.tsx:106]: @eddysims

I got this from Eddy's example, where does the magic `1.2` number come from? Should this be a dynamically retrieved value from somewhere?

File: packages/components/src/FormField/FormField.css [hunk 10]
```diff
@@ -72,7 +72,6 @@
 
 textarea.formField {
   height: auto;
-  min-height: calc(var(--field--height) * 2);
```
Comment (crpahl) [inline packages/components/src/FormField/FormField.css:75]: I'm guessing this still needs to be removed? Sorry I wasn't around when this change was made.

**Comment by crpahl** [comment]
I'm getting this weird resize behaviour and I'm not sure why. It's very slight...

resize-weirdness

**Comment by crpahl** [comment]
Not sure if this is something I should have fixed now, but I noticed there's no padding at the bottom as the scroll area starts to expand. It looks like this is an existing problem on master too?

image

**Comment by crpahl** [comment]
@seanhealy Ready for a follow-up review! I added a few comments and in-line comments for things I wasn't sure about.

File: .gitignore (hunk 6 above)
Comment (seanhealy) [reply .gitignore:9]: 😂

File: packages/components/src/FormField/FormField.css (hunk 10 above)
Comment (eddysims) [reply packages/components/src/FormField/FormField.css:75]: Im pretty sure it does, in the case you want, `min-height` should be covered by the `rows` attribute

File: packages/components/src/InputText/InputText.tsx [hunk 11]
```diff
@@ -25,9 +30,12 @@ interface MultilineProps extends BaseProps {
   readonly multiline: true;
 
   /**
-   * Specifies the visible height of a long answer form field.
+   * Specifies the visible height of a long answer form field. Can be in the
+   * form of a single number to set a static height, or an associative array
+   * with a min and max keys indicating the minimum number of visible rows, and
+   * the maximum number of visible rows.
```
Comment (eddysims) [inline packages/components/src/InputText/InputText.tsx:36]: lets change `associative array` to `object`.

File: packages/components/src/InputText/InputText.tsx (hunk 9 above)
Comment (eddysims) [reply packages/components/src/InputText/InputText.tsx:106]: yeah, this isnt the best way... we can do something like

```
const lineHeight = window
        .getComputedStyle(inputRef.current)
        .getPropertyValue("line-height");
```

*note* I grabbed this from a slack message, so it might not be exactly correct, but it should be similar to what you want.

**Comment by eddysims** [comment]
@crpahl back to you with some changes. 

One other note. because we are not taking padding/border into consideration on `resize()` we see a small shrink in the textarea on first keystroke when we have something like `rows={{min: 3, max: 4}}`. This is because we remove the bottom padding from the size calculation.

**Comment by crpahl** [comment]
@eddysims In regards to this: https://github.com/GetJobber/atlantis/pull/303#issuecomment-661245148

If you look at the gif I posted here: https://github.com/GetJobber/atlantis/pull/303#issuecomment-659699978

The height change is also happening when `resize()` isn't getting called. So something else is causing the resize and I'm not sure what it is... I know I'm missing something obvious, but I'm not sure what it is.

File: packages/components/src/InputText/InputText.tsx (hunk 9 above)
Comment (crpahl) [reply packages/components/src/InputText/InputText.tsx:106]: Ah it's the line height. Makes sense now 👍

**Comment by crpahl** [comment]
@eddysims I've addressed your changes. The weird shrinking no longer happens for this new change. Resizing still happens here though https://github.com/GetJobber/atlantis/pull/303#issuecomment-659699978 (the case where no range is given).

This seems to happen because `min-height: calc(var(--field--height) * 2);` was removed from the CSS and this seemed to be preventing the shrinking in that case.

**Comment by eddysims** [comment]
@crpahl did some digging and it looks like the vertical padding is just not even when the input changes from normal to `miniLabel`.

With no text in the `textarea` we get a total of `28px` in vertical padding

image


With text it changes to `24px` in total

image

One way I have found to fix this is to change https://github.com/GetJobber/atlantis/blob/74166e0dd3bda7324c534709e67823241499a2ff/packages/components/src/FormField/FormField.css#L157 to `--field--padding-bottom: var(--space-small);`. I _think_ that will work without changing the other inputs

**Comment by crpahl** [comment]
@eddysims Thanks, that seemed to fix it!

**Comment by crpahl** [comment]
@eddysims Back to you :) Let me know if anything else needs updating!
